lib/application.js

```diff
@@ -377,6 +377,50 @@ app.set = function set(setting, val) {
   return this;
 };
 
+/**
+ * Extend `setting` with `val`, or return `setting`'s value.
+ *
+ *    app.set('foo', { 'bar' : 1 })
+ *    app.expand('foo', { 'baz' : 1 });
+ *    app.expand('foo');
+ *    // => { "bar" : 1, "baz" : 1 }
+ *
+ * Mounted servers inherit their parent server's settings.
+ *
+ * @param {String} setting
+ * @param {*} [val]
+ * @return {Server} for chaining
+ * @public
+ */
+
+app.extend = function(setting, val) {
+  if (arguments.length === 1) {
+    // app.get(setting)
+    return this.settings[setting];
+  }
+
+  if (typeof val === 'undefined') {
+    return this;
+  } else if (val === null || typeof val !== 'object') {
+    throw new Error('value needs to be an object');
+  }
+
+  debug('extend "%s" with %o', setting, val);
+
+  var currentSetting = this.settings[setting];
+
+  // no object yet, set value
+  if (currentSetting === null || typeof currentSetting !== 'object') {
+    this.settings[setting] = val;
+    return this;
+  }
+
+  // extend value
+  this.settings[setting] = Object.assign(this.settings[setting], val);
```

Comment:
The first argument in `Object.assign` is the destination object.  So now that I read this more closely, this is probably the culprit for the parent issue, not to mention it being redundant to set the value twice.  Start by just doing this:

```
this.settings[setting] = Object.assign({}, currentSetting, val)
```

Reply:
Seems to be passing Douglas's reference tests. Depending on `currentSetting` I will also let the base object be a variable of either an array or literal.